src/Dropdown/Dropdown.tsx

```diff
@@ -22,30 +22,29 @@ export type NonDefaultProps<OptionType> = CommonProps<OptionType> &
         isClearable?: false;
         value: OptionType;
         onChange: (value: OptionType) => void;
-      });
+      }
+  );
 
 export namespace Dropdown {
   export type Props<OptionType> = NonDefaultProps<OptionType> &
     Partial<DefaultProps>;
 }
 
 export class Dropdown<OptionType> extends React.PureComponent<
-  NonDefaultProps<OptionType> & DefaultProps
+  Dropdown.Props<OptionType>
 > {
   static defaultProps = defaultProps;
 
   render() {
     const {
-      props: {
-        className,
-        components: customComponents,
-        innerRef,
-        allowCreate,
-        size,
-        flat,
-        ...props
-      }
-    } = this;
+      className,
+      components: customComponents,
+      innerRef,
+      allowCreate,
+      size,
+      flat,
+      ...props
+    } = this.props as NonDefaultProps<OptionType> & DefaultProps;
```

Comment:
added the cast as per comment above

Reply:
https://www.bookstack.cn/read/typescriptlang-3.7/spilt.58.055dccf6c3f8aae3.md

Reply:
uhm, I completely forgot about this.
Anyway, it appears to be breaking TS inference on the clearable/notClearable union. For instance [this line](https://github.com/buildo/react-components/pull/1431/files#diff-459efa0bde6087f11ee1c3da65816e1cR28) is breaking with:

```
No overload matches this call.
  Overload 1 of 2, '(props: Readonly<Pick<Props<unknown>, string | number> & { flat?: boolean | undefined; innerRef?: ((ref: StateManager<unknown, Select<unknown>> | null) => void) | undefined; options: GroupedOptionsType<...> | OptionsType<...>; } & { ...; } & { ...; } & DefaultProps> | Readonly<...> | Readonly<...> | Readonly<...>): Dropdown<...>', gave the following error.
    Type '(_: string | null) => void' is not assignable to type '((value: unknown) => void) | ((value: unknown) => void)'.
      Type '(_: string | null) => void' is not assignable to type '(value: unknown) => void'.
        Types of parameters '_' and 'value' are incompatible.
          Type 'unknown' is not assignable to type 'string | null'.
            Type 'unknown' is not assignable to type 'string'.
  Overload 2 of 2, '(props: (Pick<Props<unknown>, string | number> & { flat?: boolean | undefined; innerRef?: ((ref: StateManager<unknown, Select<unknown>> | null) => void) | undefined; options: GroupedOptionsType<...> | OptionsType<...>; } & { ...; } & { ...; } & DefaultProps) | (Pick<...> & ... 3 more ... & DefaultProps) | (Pick<...> & ... 4 more ... & DefaultProps) | (Pick<...> & ... 4 more ... & DefaultProps), context?: any): Dropdown<...>', gave the following error.
    Type '(_: string | null) => void' is not assignable to type '((value: unknown) => void) | ((value: unknown) => void)'.
      Type '(_: string | null) => void' is not assignable to type '(value: unknown) => void'.
```
(note that dtslint tests were not running before this PR because of a misconfiguration, and were already failing)

I didn't really spend much time trying to understand the real issue better 🤮

But considering that `defaultProps` is a custom react feature that solves.. 0 real problems, and is not supported nicely by TS apparently, why don't we just stop using it? 👼